### ISSUE TITLE
Climate sensor

### ISSUE THREAD

**Issue by Sonusss** [HASwitchPlate/openHASP#230]
It would be nice if we could add climate sensors like DHT22 or similar...

**Comment by fvanroie** [comment]
Digging through my sensors bin, I don't seem to have any of these sensors. I'll order some from aliexpress but it's going to take a while to ship...

I might get my hands on a couple of Dallas sensors sooner though.

**Comment by EnricoSanchez** [comment]
First of all thanks for this great project.  But could you please add the bme280 sensor for heating control.  So I don't need any additional hardware to use the display better for smart room control

**Comment by fvanroie** [comment]
Adding a sensor not the hard part. But to what endpoint/topic does the sensor data need to be sent and in what format?
On the receiving end there has to be a controller or some logic that processes these messages, like the Custom Component.

If I know what the message requirements are, it will be a lot easier to provide the data in a way that can be processed.

**Comment by Qargh** [comment]
(Not) One data model fits them all.

I'm sure if I ask nine people in wich form the collected data of a bme280 for example should be transmitted i will get ten answers :-).
The Temperature for example:

- temp: "19.6 °C" -> will be nice to show it on a screen but for math you need to parse the value first.
- temp: 19,600 -> great for most european math. In USA ... oops ... you'll need a lot of ice. And by the way. °C or °F?
- temperature: 19.6, tempUnit: "Celsius" -> show on screen as it is ... no way ... but concatenating and localising may be easier then parsing ... hmmm

Going further:

- how should the data be structured if you have .. lets say two temp units to measure a delta?
  So what suggestion for a format is the right one?
- People out there all have their own preferences which sensors or other hardware they want to attach and what automatic behavior they need. If you take a short look at the Tasmota Firmware or ESP Easy you can see which dimensions the support of devices and ruled behavior can reach easily.

Lets see some more docs:

> - [Amazon Alexa](https://developer.amazon.com/en-US/docs/alexa/device-apis/alexa-interface.html)
> - [Google Smarthome](https://developers.google.com/assistant/smarthome/overview)
> - [EnOcean EEP](https://www.enocean-alliance.org/wp-content/uploads/2017/05/EnOcean_Equipment_Profiles_EEP_v2.6.7_public.pdf)
> - [Tasmota firmware](https://tasmota.github.io/docs/)
> - [ESP Easy - BME280](https://www.letscontrolit.com/wiki/index.php/BME280)

**Wow, keep this horse in check!**

As I remember there are several requests (Issues and Discussions) to support alternative hardware within OpenHASP the last 2 years. But I think it will take some "glue" between the CUSTOM Interface and the differnt hardware implementations to get that managed.

Maybe:
- some more advanced examples on howto get managed a build and deploy pipeline for the firmware.
- some examples on how to create a "controller" on other Smart Home Systems (e.g. openHAB).
- some guidelines for the "glue".
- it's all there and i don't know the www links (if true, please tell me)

But to end the philosophizing. This is my suggestion for the data format.

```json
{
  "custom": {
    "<topic compliant string>": {
      "customId": "<topic compliant string>",
      "meta": {
        "deviceType": "eg. BME208",
        "pressureUnit": "hPa",
        "tempUnit": "C"
      },
      "properties": {
        "lastUpdate": "<iso8601 utc zulu time>",
        "temperature": 21.7,
        "humidity": 66.6,
        "pressure": 988.6
      }
    }
  }
}
```

Remarks:
- Just read the json properties in chain as the custom topic (eg. /custom/CUSTOM_ID/properties/temperature)
- populate meta less more then properties and not in the "get sensor data" callback
- don't use an array's in the "custom" property or numeric indexes. Stay with the custom_id. It will make trouble if you want to do things like a JSON Merge or JSON Patch on more advanced "controllers" in the backend. And, be honest, what is "/plate_46/state/custom/8/xyz"?

Regards to all the lovely work on OpenHASP.

**Comment by EnricoSanchez** [comment]
> Adding a sensor not the hard part. But to what endpoint/topic does the sensor data need to be sent and in what format? On the receiving end there has to be a controller or some logic that processes these messages, like the Custom Component.
> 
> If I know what the message requirements are, it will be a lot easier to provide the data in a way that can be processed.

For me, it would bei enough when in see im my mqtt Broker (i use iobroker with mqtt Adapter) an additonal datapoint, like for example a Display Button, but with the temperature , so i can use the state to control my heating valve and also send it back to the Display

**Comment by Janverhu** [comment]
I just finished my_custom.ccp which features a motion sensor and a dht22 temp/humidity sensor. I'm fairly new to c++, arduino and openHasp so comments are welcome. I thought some folks here would be interested...
[my_custom .cpp.txt](https://github.com/HASwitchPlate/openHASP/files/8054735/my_custom.cpp.txt)

**Comment by Humancell** [comment]
There some other critical aspects to adding sensors into a device. You really must also include an identifier of exactly what part/chip/transducer is providing the data so that you can know the accuracy and range of the sensor.

We often think that getting the temperature or humidity is enough, but without understanding the accuracy or the part or where the reading is within it's operational range, your code could be making bad choices in controlling things.

**Comment by Janverhu** [comment]
Sure, for my application which is not a critical one it suffices to double check the numbers with another device I know is accurate. I had to change my design a couple of times and ended up with an external case for the sensor because the board gives off too much heat. But that’s the responsibility of the diy’er to do…

**Comment by fantasmisiciliani** [comment]
@fvanroie I added the arduino-sht library in platform-override.ini and wrote few lines of code in my_custom.cpp to read from an SHT31 sensor, but I found that Wire (first I2C port) is already initialized (Wire1 is used by the touchscreen) and I had to call Wire.end() before re-initializing it with my custom pins. In this way it works, but is there something I broke?
I have a WT32-SC01 Plus with 3ab88a3.
